{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data input - Option 1 - Generate random data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate data frame of random numbers with artificial ENSEMBL gene ID\n",
    "np.seed = 1\n",
    "number_of_rows = 30000\n",
    "expression_df = pd.DataFrame({\n",
    "    \"ENSEMBL_Gene_ID\": [f\"ENSG{x:010}\" for x in range(1, number_of_rows + 1)],\n",
    "    \"fold_change\": np.random.randn(number_of_rows)* 5,\n",
    "    \"adjusted_p_value\": np.random.randint(0, 10000, number_of_rows)/10000,\n",
    "    \"gene_length\": np.abs(np.round((np.random.randn(number_of_rows)* 50000).astype(int), 0)),\n",
    "    \"chromosome\": [f\"Chr{chr_number}\" for chr_number in np.random.choice(range(1, 24), size=number_of_rows)]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSEMBL_Gene_ID</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>adjusted_p_value</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG0000000001</td>\n",
       "      <td>-3.093222</td>\n",
       "      <td>0.9598</td>\n",
       "      <td>109109</td>\n",
       "      <td>Chr5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG0000000002</td>\n",
       "      <td>5.079444</td>\n",
       "      <td>0.0641</td>\n",
       "      <td>42231</td>\n",
       "      <td>Chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG0000000003</td>\n",
       "      <td>0.361165</td>\n",
       "      <td>0.0244</td>\n",
       "      <td>47592</td>\n",
       "      <td>Chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG0000000004</td>\n",
       "      <td>-6.597946</td>\n",
       "      <td>0.8434</td>\n",
       "      <td>94922</td>\n",
       "      <td>Chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG0000000005</td>\n",
       "      <td>-0.775996</td>\n",
       "      <td>0.0339</td>\n",
       "      <td>53489</td>\n",
       "      <td>Chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG0000000006</td>\n",
       "      <td>0.164095</td>\n",
       "      <td>0.8092</td>\n",
       "      <td>17486</td>\n",
       "      <td>Chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG0000000007</td>\n",
       "      <td>9.507698</td>\n",
       "      <td>0.7899</td>\n",
       "      <td>28534</td>\n",
       "      <td>Chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG0000000008</td>\n",
       "      <td>-0.229587</td>\n",
       "      <td>0.9606</td>\n",
       "      <td>12830</td>\n",
       "      <td>Chr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG0000000009</td>\n",
       "      <td>15.503495</td>\n",
       "      <td>0.3538</td>\n",
       "      <td>52764</td>\n",
       "      <td>Chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG0000000010</td>\n",
       "      <td>-2.061291</td>\n",
       "      <td>0.8111</td>\n",
       "      <td>23617</td>\n",
       "      <td>Chr4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ENSEMBL_Gene_ID  fold_change  adjusted_p_value  gene_length chromosome\n",
       "0  ENSG0000000001    -3.093222            0.9598       109109       Chr5\n",
       "1  ENSG0000000002     5.079444            0.0641        42231       Chr3\n",
       "2  ENSG0000000003     0.361165            0.0244        47592       Chr7\n",
       "3  ENSG0000000004    -6.597946            0.8434        94922      Chr16\n",
       "4  ENSG0000000005    -0.775996            0.0339        53489       Chr8\n",
       "5  ENSG0000000006     0.164095            0.8092        17486       Chr9\n",
       "6  ENSG0000000007     9.507698            0.7899        28534       Chr8\n",
       "7  ENSG0000000008    -0.229587            0.9606        12830       Chr6\n",
       "8  ENSG0000000009    15.503495            0.3538        52764      Chr10\n",
       "9  ENSG0000000010    -2.061291            0.8111        23617       Chr4"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_fh = open(\"Expression_data.csv\", \"w\")\n",
    "output_fh.write(\"# Randomly generate gene expression data\\n\")\n",
    "expression_df.to_csv(output_fh, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data input - Option 2 - Download (randomly generated) data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "source_url = \"https://github.com/foerstner-lab/Bits_and_pieces_for_the_carpentries_workshops/blob/main/python/Expression_data.csv?raw=true\"\n",
    "expression_data_file = \"Expression_data.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_df = pd.read_csv(expression_data_file, sep=\"\\t\", comment=\"#\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The actual filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSEMBL_Gene_ID</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>adjusted_p_value</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG0000000001</td>\n",
       "      <td>-3.093222</td>\n",
       "      <td>0.9598</td>\n",
       "      <td>109109</td>\n",
       "      <td>Chr5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG0000000002</td>\n",
       "      <td>5.079444</td>\n",
       "      <td>0.0641</td>\n",
       "      <td>42231</td>\n",
       "      <td>Chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG0000000003</td>\n",
       "      <td>0.361165</td>\n",
       "      <td>0.0244</td>\n",
       "      <td>47592</td>\n",
       "      <td>Chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG0000000004</td>\n",
       "      <td>-6.597946</td>\n",
       "      <td>0.8434</td>\n",
       "      <td>94922</td>\n",
       "      <td>Chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG0000000005</td>\n",
       "      <td>-0.775996</td>\n",
       "      <td>0.0339</td>\n",
       "      <td>53489</td>\n",
       "      <td>Chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29939</th>\n",
       "      <td>ENSG0000029940</td>\n",
       "      <td>-1.977256</td>\n",
       "      <td>0.6678</td>\n",
       "      <td>959</td>\n",
       "      <td>Chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29940</th>\n",
       "      <td>ENSG0000029941</td>\n",
       "      <td>1.229374</td>\n",
       "      <td>0.1838</td>\n",
       "      <td>123072</td>\n",
       "      <td>Chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29941</th>\n",
       "      <td>ENSG0000029942</td>\n",
       "      <td>2.232070</td>\n",
       "      <td>0.7865</td>\n",
       "      <td>49962</td>\n",
       "      <td>Chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29942</th>\n",
       "      <td>ENSG0000029943</td>\n",
       "      <td>3.716701</td>\n",
       "      <td>0.1706</td>\n",
       "      <td>39327</td>\n",
       "      <td>Chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29943</th>\n",
       "      <td>ENSG0000029944</td>\n",
       "      <td>-0.737347</td>\n",
       "      <td>0.1022</td>\n",
       "      <td>37070</td>\n",
       "      <td>Chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29944 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ENSEMBL_Gene_ID  fold_change  adjusted_p_value  gene_length chromosome\n",
       "0      ENSG0000000001    -3.093222            0.9598       109109       Chr5\n",
       "1      ENSG0000000002     5.079444            0.0641        42231       Chr3\n",
       "2      ENSG0000000003     0.361165            0.0244        47592       Chr7\n",
       "3      ENSG0000000004    -6.597946            0.8434        94922      Chr16\n",
       "4      ENSG0000000005    -0.775996            0.0339        53489       Chr8\n",
       "...               ...          ...               ...          ...        ...\n",
       "29939  ENSG0000029940    -1.977256            0.6678          959       Chr7\n",
       "29940  ENSG0000029941     1.229374            0.1838       123072       Chr4\n",
       "29941  ENSG0000029942     2.232070            0.7865        49962       Chr3\n",
       "29942  ENSG0000029943     3.716701            0.1706        39327      Chr16\n",
       "29943  ENSG0000029944    -0.737347            0.1022        37070       Chr1\n",
       "\n",
       "[29944 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fold_change</th>\n",
       "      <th>adjusted_p_value</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>29944.000000</td>\n",
       "      <td>29944.000000</td>\n",
       "      <td>29944.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.004812</td>\n",
       "      <td>0.502080</td>\n",
       "      <td>40053.986942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.009114</td>\n",
       "      <td>0.289636</td>\n",
       "      <td>30281.730134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-18.546675</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>-3.377455</td>\n",
       "      <td>0.248875</td>\n",
       "      <td>15990.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.018451</td>\n",
       "      <td>0.503500</td>\n",
       "      <td>33806.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>3.426894</td>\n",
       "      <td>0.756200</td>\n",
       "      <td>57709.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>19.805340</td>\n",
       "      <td>0.999900</td>\n",
       "      <td>199697.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        fold_change  adjusted_p_value    gene_length\n",
       "count  29944.000000      29944.000000   29944.000000\n",
       "mean       0.004812          0.502080   40053.986942\n",
       "std        5.009114          0.289636   30281.730134\n",
       "min      -18.546675          0.000000       0.000000\n",
       "25%       -3.377455          0.248875   15990.500000\n",
       "50%        0.018451          0.503500   33806.500000\n",
       "75%        3.426894          0.756200   57709.500000\n",
       "max       19.805340          0.999900  199697.000000"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSEMBL_Gene_ID</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>adjusted_p_value</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG0000000005</td>\n",
       "      <td>-0.775996</td>\n",
       "      <td>0.0339</td>\n",
       "      <td>53489</td>\n",
       "      <td>Chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG0000000006</td>\n",
       "      <td>0.164095</td>\n",
       "      <td>0.8092</td>\n",
       "      <td>17486</td>\n",
       "      <td>Chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG0000000007</td>\n",
       "      <td>9.507698</td>\n",
       "      <td>0.7899</td>\n",
       "      <td>28534</td>\n",
       "      <td>Chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG0000000008</td>\n",
       "      <td>-0.229587</td>\n",
       "      <td>0.9606</td>\n",
       "      <td>12830</td>\n",
       "      <td>Chr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG0000000009</td>\n",
       "      <td>15.503495</td>\n",
       "      <td>0.3538</td>\n",
       "      <td>52764</td>\n",
       "      <td>Chr10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ENSEMBL_Gene_ID  fold_change  adjusted_p_value  gene_length chromosome\n",
       "4  ENSG0000000005    -0.775996            0.0339        53489       Chr8\n",
       "5  ENSG0000000006     0.164095            0.8092        17486       Chr9\n",
       "6  ENSG0000000007     9.507698            0.7899        28534       Chr8\n",
       "7  ENSG0000000008    -0.229587            0.9606        12830       Chr6\n",
       "8  ENSG0000000009    15.503495            0.3538        52764      Chr10"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ids = [\"ENSG0000000005\",\"ENSG0000000006\",\"ENSG0000000007\",\"ENSG0000000008\",\"ENSG0000000009\"]\n",
    "selection_by_gene_name = expression_df[expression_df[\"ENSEMBL_Gene_ID\"].isin(gene_ids)]\n",
    "selection_by_gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to file\n",
    "selection_by_gene_name.to_csv(\"Expression_data_selection.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'adjusted_p_values'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/nix/store/7q89lh15ykx8pzf81m4qpvkq613rcbd5-python3.9-pandas-1.2.4/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3079\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3080\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'adjusted_p_values'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-d774d114776b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mselection_by_adj_p_value\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexpression_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mexpression_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"adjusted_p_values\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m0.05\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/nix/store/7q89lh15ykx8pzf81m4qpvkq613rcbd5-python3.9-pandas-1.2.4/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3022\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3023\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3024\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3025\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3026\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nix/store/7q89lh15ykx8pzf81m4qpvkq613rcbd5-python3.9-pandas-1.2.4/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3080\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3081\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3082\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3084\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'adjusted_p_values'"
     ]
    }
   ],
   "source": [
    "selection_by_adj_p_value = expression_df[expression_df[\"adjusted_p_values\"] < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selection_by_adj_p_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_df.fold_change.hist()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
